### ISSUE TITLE
Email Invitations

### ISSUE THREAD

**Issue by philippneugebauer** [ckulka/baikal-docker#31]
Have you thought about solutions for enabling email invitations?

When I try that, I get the following error:
```
sh: 1: /usr/sbin/sendmail: not found
```

**Comment by ckulka** [comment]
I figure installing sendmail is all that's needed (https://stackoverflow.com/questions/38299776/usr-sbin-sendmail-not-found/38299891). I'll post an update when I updated the Dockerfile, would you be able to test it then?

**Comment by philippneugebauer** [comment]
yeah, for sure. I tried it and got stuck with the postfix installation :(

**Comment by ckulka** [comment]
Ok, it's building now - based on previous Docker Hub builds, that'll take about 1.5h. Can you then please give it a try with one of the following (well, ideally both if you can) images?

* `ckulka/baikal:experimental-apache`
* `ckulka/baikal:experimental-nginx`

**Comment by philippneugebauer** [comment]
Yeah, I found that SO answer, too. I am still not sure, if that's the right thing to do since that means to really host such an smtp server which comes with quite some effort. I think it would be nicer to be able to include a service or at least having it configured outside the baikal container?

**Comment by philippneugebauer** [comment]
that description might be interesting for the setup: https://mchodled.com/posts/2020/Apr/03/de-google-your-calendar-with-baikal/#configure-email-invitation-plugin

**Comment by philippneugebauer** [comment]
I will provide a PR with my changes soon. It's working but I struggle to set the values in `/etc/msmtprc` with ENVs

**Comment by ahgraber** [comment]
@philippneugebauer I'm working on this myself.  

Using the solution from the mchodled.com link, I think it's possible to mount the files as volumes, so the end user should just have an `msmtprc` or `msmtp_php` locally with their config, and it is mounted in the docker-compose:

```yaml
#...
services:
  baikal:
    image: ckulka/baikal:nginx
    container_name: baikal
    restart: unless-stopped
    volumes:
      - ./config:/var/www/baikal/config
      - ./data:/var/www/baikal/Specific
      - ./msmtp_php:/etc/.msmtp_php
```

The real trick is to update the `sendmail_path` directly, and I think you'd have to rely on some bash scripting in the entrypoint/cmd script`start.sh`.  You could do something like:

```bash
sed -i 's/;sendmail_path =/sendmail_path = "/usr/bin/msmtp -C /etc/.msmtp_php --logfile /var/log/msmtp.log -a ${SENDMAIL_ACCOUNT} -t"/' /etc/php/8.0/fpm/php.ini
```
where SENDMAIL_ACCOUNT is an environmental variable

I haven't had time to test yet, though.

**Comment by philippneugebauer** [comment]
have you had look at my PR? @ahgraber

**Comment by ahgraber** [comment]
Apologies, was moving fast trying to solve the same problem and just saw the comment about struggling to set values -- didn't see the PR link.  Wasn't trying to be pedantic, just wanted to help :)

Your solution is specific to the :apache version?

**Comment by philippneugebauer** [comment]
I did not mean it like that, I rather wanted to point out the PR to get some feedback from you

**Comment by ahgraber** [comment]
I'll follow up with comments on the PR itself.

**Comment by lennvilardi** [comment]
Is this problem closed ? Is there a tutorial to use email invites ?

**Comment by philippneugebauer** [comment]
there is a possible option in https://github.com/ckulka/baikal-docker/pull/32

**Comment by MrPinrel** [comment]
I recently installed this container (great work!) on my Synology NAS and would love to get the email invites working. How would I go about enabling this? Do I need to download the experimental nginx version? Or is it going to get included in the latest release soon?

Happy to be a tester but would need some help on how to get started. Also, I use a mail relay, not sure if this makes a difference?

**Comment by philippneugebauer** [comment]
build the dockerfile from #32 on your own and use it

**Comment by MrPinrel** [comment]
> build the dockerfile from #32 on your own and use it

Thanks, I got it to build but it is complaining that msmtprc is not installed in the synology box. I found msmtprc and tried to install it, but it tells me that there is no suitable C compiler found to do that....now trying to figure out how to install that. Let me know if you have any ideas.

For a broader distribution it might be good to include those steps in the automated installation.

**Comment by MrPinrel** [comment]
update, I'm not having any luck getting msmtprc to build in my synology box. I get errors about libraries missing and errors in the MAKE file. Does anybody know where I could find the already built images? It is a DS918+ with the Intel Celeron processor.

**Comment by lennvilardi** [comment]
Seriously considering installing Nextcloud just to get email notification, it's too complicated to compile msmtp for a DS918+... is there another way (simplier?) ?

**Comment by MrPinrel** [comment]
> Seriously considering installing Nextcloud just to get email notification, it's too complicated to compile msmtp for a DS918+... is there another way (simplier?) ?

Lenn thanks for this note, it caused me to look into installing nextcloud to a caldav server with working invites running on DSM 7.0. I followed[ these instructions](https://mariushosting.com/how-to-install-nextcloud-on-your-synology-nas/) and it was really easy. First I tried to install in Docker but then I decided to do the native install.

**Comment by vaskozl** [comment]
I've been a happy radicale user for a few years now, after initially migrating away a couple of years ago. The reason was wanting to be able to send calendar invites for personal events, e.g. holiday flights.

I've managed to use sendmail as included by the experimental images but that required a number of hacks:

* One obviously needs to the sendmail daemon.

* The first problem one might encounter is that sendmail needs a FQDN in the hosts file, otherwise it hangs for a minute every time a mail is sent. (and the daemon is started). One needs to add to the hosts file. You can't even use `sed -i` directly due to the hosts file being mounted (and not moveable).
* Your mail wil literally not be accepted by anyone unless you have a proper mail server e.g. postfix somewhere on the network. You can do this by adding a SMART_HOST to sendmail.mc
* Even then, your emails may still go to spam, because the `mail()` function as used by postfix doesn't set the envelope sender (via the  `-f`flag)

This is the command I'm using that actually delivers mail properly to gmail and the like:
```
              mv /usr/sbin/sendmail /usr/sbin/sendmail-og && 
              echo '#!/bin/sh' > /usr/sbin/sendmail && chmod +x /usr/sbin/sendmail &&  
              echo '/usr/sbin/sendmail-og -f baikal@sko.ai $@' >> /usr/sbin/sendmail &&
              echo 'nginx' >> /etc/mail/trusted-users &&
              echo "define(\`SMART_HOST', \`mail.yourlocal.domain')dnl" >> /etc/mail/sendmail.mc &&
              cp /etc/hosts /etc/hosts.new && 
              sed -i "s/$HOSTNAME/$HOSTNAME baikal.yourlocal.domain/" /etc/hosts.new && 
              cp /etc/hosts.new /etc/hosts && 
              service sendmail start && 
              /etc/init.d/php8.0-fpm start && 
              chown -R nginx:nginx /var/www/baikal/Specific && 
              nginx -g "daemon off;"
```

If any one is interested in the k8s manifest I used:
https://github.com/vaskozl/home-infra/commit/9bbdf3d611085edff0309fc0fcfe63cda2e73f1e

**Comment by ckulka** [comment]
Thanks all, especially @philippneugebauer and @vaskozl. My proposal below is basically just a mix of #32 and https://github.com/ckulka/baikal-docker/issues/31#issuecomment-1008344665.

Would like to hear what you think of this:

1. The Docker image won't contain a mail server, it will rely on an external one (another container or a service like Gmail)

2. The container will use sendmail and not msmtp

   Seems like sendmail is more widely adopted (<https://trends.google.de/trends/explore?q=sendmail,msmtp>), looks like being more of the default option to go with, and judging from the releases both are about equally frequently maintained.

   It also seems easier to use mounted secrets with sendmail compared to msmtp.

3. Sendmail configuration will be applied during container start using three environment variables: the host, the "from" email address, and the additional `sendmail.mc` configuration

The `docker-compose.yaml` file would look like this:

```yaml
version: '2'
services:
  baikal:
    image: ckulka/baikal:nginx
    environment:
      BAIKAL_SERVERNAME: baikal.yourdomain.com
      SENDMAIL_FROM: baikal@yourdomain.com
      SENDMAIL_MAILER_DEFINITIONS: |
        define(\`SMART_HOST', \`mail.yourdomain.com')dnl
```

The `SENDMAIL_FROM` environment variable with a prefixed `-f` flag is added to `mail.force_extra_parameters` in `/etc/php/8.1/fpm/php.ini`, so that we don't have to customize the `/usr/sbin/sendmail` script.

The `SENDMAIL_MAILER_DEFINITIONS` environment variable is inserted into `/etc/mail/sendmail.mc` underneath `MAILER_DEFINITIONS`, according to [Configure Sendmail](https://linuxconfig.org/configuring-gmail-as-sendmail-email-relay) it's important that it's right underneath it ¯\ _ (ツ)_/¯. It can have multiple lines to allow for more configuration like authentication.

The `BAIKAL_SERVERNAME` is inserted to `/etc/hosts` just like @vaskozl has it in the script. **What I don't know/my question:** is this the same as the FQDN where Baikal is accessed, or could this be a different one?

If it's always the same one, we can reuse `BAIKAL_SERVERNAME` which is already a (somewhat) required parameter for the Apache image. If it can be different, I'd opt for a different variable name.
